tapdb/multiverse.go

```diff
@@ -64,21 +73,236 @@ type BatchedMultiverse interface {
 	BatchedTx[BaseMultiverseStore]
 }
 
+// ProofKey is used to uniquely identify a proof within a universe. This is
+// used for the LRU cache for the proofs themselves, which are considered to be
+// immutable.
+type ProofKey [32]byte
+
+// NewProofKey takes a universe identifier and leaf key, and returns a proof
+// key.
+func NewProofKey(id universe.Identifier, key universe.LeafKey) ProofKey {
+	idBytes := id.Bytes()
+	leafKeyBytes := key.UniverseKey()
+
+	// The proof key maps down the ID and the leaf key into a single
+	// 32-byte value: sha256(id || leaf_key)..
+	h := sha256.New()
+	h.Write(idBytes[:])
+	h.Write(leafKeyBytes[:])
+
+	return fn.ToArray[ProofKey](h.Sum(nil))
+}
+
+// numCachedProofs is the number of universe proofs we'll cache.
+const numCachedProofs = 25_000
+
+// cachedProof is a single cached proof.
+type cachedProof []*universe.Proof
+
+// Size just returns 1 as we're limiting based on the total number of proofs.
+func (c *cachedProof) Size() (uint64, error) {
+	return 1, nil
+}
+
+// leafProofCache is used to cache proofs for issuance leaves for assets w/o a
+// group key.
+type leafProofCache = *lru.Cache[ProofKey, *cachedProof]
+
+// newLeafCache creates a new leaf proof cache.
+func newLeafCache() leafProofCache {
+	return lru.NewCache[ProofKey, *cachedProof](
+		numCachedProofs,
+	)
+}
+
+// treeID is used to uniquely identify a multiverse tree.
+type treeID string
+
+// proofCache a map of proof caches for each proof type.
+type proofCache lnutils.SyncMap[treeID, leafProofCache]
+
+// newProofCache creates a new proof cache.
+func newProofCache() *proofCache {
+	return &proofCache{}
+}
+
+// fetchProof reads the cached proof for the given ID and leaf key.
+func (p *proofCache) fetchProof(id universe.Identifier,
+	leafKey universe.LeafKey) []*universe.Proof {
+
+	// First, get the sub-cache for this universe ID from the map of
+	// caches.
+	idStr := treeID(id.String())
```

Comment:
It's just a type def to make the types a bit more readable. a `treeID` is just a string. We can't use the full `universe.Identifier` here as it has a pointer, so that can mess up map key comparison.